benchmarks/low_level.cpp

```diff
@@ -0,0 +1,455 @@
+
+#include <benchmark/benchmark.h>
+#include <filesystem>
+#include <gtest/gtest.h>
+
+class low_level : public ::benchmark::Fixture {
+  protected:
+    void SetUp(const ::benchmark::State&) override {
+        auto mint_tx1 = wallet1.mint_new_coins(2, 100);
+        wallet1.confirm_transaction(mint_tx1);
+        auto mint_tx2 = wallet2.mint_new_coins(1, 100);
+        wallet2.confirm_transaction(mint_tx2);
+
+        m_of.open(m_BENCHMARK_File);
+    }
+
+    void TearDown(const ::benchmark::State&) override {
+        m_if.close();
+        m_of.close();
+        std::filesystem::remove_all(m_BENCHMARK_File);
+    }
+
+    cbdc::transaction::wallet wallet1;
+    cbdc::transaction::wallet wallet2;
+
+    cbdc::transaction::full_tx m_valid_tx{};
+    cbdc::transaction::full_tx m_valid_tx_multi_inp{};
+
+    // std::unique_ptr<secp256k1_context, decltype(&secp256k1_context_destroy)>
+    //     m_secp{secp256k1_context_create(SECP256K1_CONTEXT_SIGN
+    //                                     | SECP256K1_CONTEXT_VERIFY),
+    //            &secp256k1_context_destroy};
+    // cbdc::privkey_t m_priv0{cbdc::hash_from_hex(
+    //     "0000000000000001000000000000000000000000000000000000000000000000")};
+    // cbdc::pubkey_t m_pub0{cbdc::pubkey_from_privkey(m_priv0, m_secp.get())};
+    // cbdc::privkey_t m_priv1{cbdc::hash_from_hex(
+    //     "1000000000000001000000000000000000000000000000000000000000000000")};
+    // cbdc::pubkey_t m_pub1{cbdc::pubkey_from_privkey(m_priv1, m_secp.get())};
+    // std::unordered_set<cbdc::pubkey_t, cbdc::hashing::null> m_pubkeys{m_pub0,
+    //                                                                   m_pub1};
+
+    // for serialization
+    std::ifstream m_if;
+    std::ofstream m_of;
+
+    cbdc::istream_serializer m_is{m_if};
+    cbdc::ostream_serializer m_os{m_of};
+
+    static constexpr auto m_BENCHMARK_File
+        = "serial_BENCHMARK_File.dat";
+};
+
+// create_tx 1-1
+BENCHMARK_F(low_level, create_1to1_tx)(benchmark::State& state) {
+    // if(state.thread_index() == 0) {
+    //     GTEST_LOG_(INFO) << "Starting transaction benchmarks ... ";
+    // }
+    for(auto _ : state) {
+        m_valid_tx
+            = wallet1.send_to(100, wallet2.generate_key(), true).value();
+        state.PauseTiming();
+        wallet1.confirm_transaction(m_valid_tx);
+        wallet2.confirm_transaction(m_valid_tx);
+        // send back for reuse
+        m_valid_tx
+            = wallet2.send_to(100, wallet1.generate_key(), true).value();
+        wallet1.confirm_transaction(m_valid_tx);
+        wallet2.confirm_transaction(m_valid_tx);
+        state.ResumeTiming();
+    }
+}
+
+// create_tx 1-2
+BENCHMARK_F(low_level, create_1to2_tx)(benchmark::State& state) {
+    for(auto _ : state) {
+        state.PauseTiming();
+        cbdc::transaction::wallet wallet_a;
+        cbdc::transaction::wallet wallet_b;
+        auto mint_tx_a = wallet_a.mint_new_coins(1, 50);
```

Comment:
The `SetUp` already mints coins for the wallets at the beginning of each benchmark; are more needed?

What we're actually timing here will be clearer if each individual benchmark can minimize any other ceremony (there will probably always be some small amount, but I think less here is possible).

Reply:
To exclusively rely on the `SetUp` method would require the wallets to be initialized to hold a huge amount of value. This is because since the number of times that the benchmark is run is dynamically determined (that is, the number of times the code within the `for(auto _ : state)` block). `SetUp` does not get called before each loop, so the possible workarounds I've identified are to instantiate new wallets each loop and discard the setup time, mint new coins each loop (not sure if I should reset the wallets in this scenario), or force a maximum number of loops. 

Not sure which method would be most beneficial here, happy to hear any thoughts.

Reply:
Mmm, that's interesting, and kind of an odd limitation. In that case, I'd suggest making an additional helper function that you call on each iteration (outside timing) to setup the wallets the same way each time. That will still reduce the noise in each benchmark definition.

Reply:
Resolving because of modifications made